### PR TITLE
[jit] Attribute serialization improvements

### PR DESCRIPTION
**Pull request by driazati**
* adds attributes to `ScriptModule.__getattr__` so they can be accessed in Python after re-importing
* full support for all the possible values for an `int64_t`
    * this necessitated a bunch more `pushWhatever` functions, so re-introduced a templated version to cut down on duplicate code
* tests to validate references / value sharing works
* adds `torch.jit.Unpickler` which people can use to de-serialize the pickle files into Python / have a quick reference on how to do this without PyTorch

### CODE REVIEW

File: torch/jit/_pickle.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    def __setstate__(self, id):
+        self.id = id
+
+
+class IntList(object):
+    def __setstate__(self, data):
+        self.data = data
+
+
+class LiteralTensor(object):
```
Comment (zdevito) [inline torch/jit/_pickle.py]: This seems unused in this PR. Hard to know if it would work or not. Best to add it later.

**Review by zdevito — APPROVED** [review]

**Comment by ezyang** [comment]
This appears to have broken Windows. 

```
21:43:09 ======================================================================
21:43:09 ERROR: test_serialization_sharing (__main__.TestScript)
21:43:09 ----------------------------------------------------------------------
21:43:09 Traceback (most recent call last):
21:43:09   File "test_jit.py", line 97, in TemporaryFileName
21:43:09     yield f.name
21:43:09   File "test_jit.py", line 10604, in test_serialization_sharing
21:43:09     pickled_data = archive.read(os.path.join(archive_name, 'attributes.pkl'))
21:43:09   File "C:\Jenkins\Miniconda3\lib\zipfile.py", line 1337, in read
21:43:09     with self.open(name, "r", pwd) as fp:
21:43:09   File "C:\Jenkins\Miniconda3\lib\zipfile.py", line 1375, in open
21:43:09     zinfo = self.getinfo(name)
21:43:09   File "C:\Jenkins\Miniconda3\lib\zipfile.py", line 1304, in getinfo
21:43:09     'There is no item named %r in the archive' % name)
21:43:09 KeyError: "There is no item named 'tmp536fannv\\\\attributes.pkl' in the archive"
21:43:09 
21:43:09 During handling of the above exception, another exception occurred:
21:43:09 
21:43:09 Traceback (most recent call last):
21:43:09   File "test_jit.py", line 10613, in test_serialization_sharing
21:43:09     .check_count("BINGET", 2, exactly=True).run(out.getvalue())
21:43:09   File "C:\Jenkins\Miniconda3\lib\contextlib.py", line 99, in __exit__
21:43:09     self.gen.throw(type, value, traceback)
21:43:09   File "test_jit.py", line 99, in TemporaryFileName
21:43:09     os.unlink(f.name)
21:43:09 PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Windows\\TEMP\\tmp536fannv'
21:43:09 
```

**Comment by ezyang** [comment]
Windows has more strict file deletion requirements; it won't let you unlink a file while you have it open.

**Comment by ezyang** [comment]
I can't unland this, there are merge conflicts.

**Comment by driazati** [comment]
@ezyang #18667 should fix the test

**Pull request closed without merging**